Monitor query readme (#16791)
Azure/azure-sdk-for-js: sdk/monitor/monitor-query/README.md

}
```

### Batch logs query

The following example demonstrates sending multiple queries at the same time using batch query API. The queries can be represented as a list of `BatchQuery` objects.

> ```

Reviewer: ```suggestion
```javascript
```

export async function main() {
  if (!monitorWorkspaceId) {
    throw new Error("MONITOR_WORKSPACE_ID must be set in the environment for this sample");
  }

  const tokenCredential = new DefaultAzureCredential();
  const logsQueryClient = new LogsQueryClient(tokenCredential);

  const queriesBatch: BatchQuery[] = [
    {
      workspaceId: monitorWorkspaceId,
      query: "AppEvents | project TimeGenerated, Name, AppRoleInstance | limit 1",
      timespan: "P1D"
    },
    {
      workspaceId: monitorWorkspaceId,
      query: "AzureActivity | summarize count()",
      timespan: "PT1H"
    },
    {
      workspaceId: monitorWorkspaceId,
      query:

>         "AppRequests | take 10  | summarize avgRequestDuration=avg(DurationMs) by bin(TimeGenerated, 10m), _ResourceId",

Reviewer: ```suggestion
        "AppRequests | take 10 | summarize avgRequestDuration=avg(DurationMs) by bin(TimeGenerated, 10m), _ResourceId",
```

      timespan: "PT1H"
    },
    {
      workspaceId: monitorWorkspaceId,
      query: "AppRequests | take 2",
      timespan: "PT1H"
    }
  ];

  const result = await logsQueryClient.queryLogsBatch(
    {
      queries: queriesBatch
    }
  );

  if (result.results == null) {
    throw new Error("No response for query");
  }

  let i = 0;
  for (const response of result.results) {
    console.log(`Results for query with id: ${response.id}`);

    if (response.error) {

>       console.log(` Query had errors:`, response.error);

Reviewer: ```suggestion
      console.log(`Query had errors:`, response.error);
```

    } else {
      if (response.tables == null) {
        console.log(`No results for query`);
      } else {
        console.log(
          `Printing results from query '${queriesBatch[i].query}' for '${queriesBatch[i].timespan}'`
        );

        for (const table of response.tables) {
          const columnHeaderString = table.columns
            .map((column) => `${column.name}(${column.type}) `)
            .join("| ");
          console.log(columnHeaderString);

          for (const row of table.rows) {
            const columnValuesString = row.map((columnValue) => `'${columnValue}' `).join("| ");
            console.log(columnValuesString);
          }
        }
      }
    }
    // next query
    i++;
  }
}
```

### Query metrics

> The following example gets metrics for a Metrics Advisor subscription. The resource URI is that of a metrics advisor resource.

Reviewer: ```suggestion
The following example gets metrics for an [Azure Metrics Advisor](https://docs.microsoft.com/azure/applied-ai-services/metrics-advisor/overview) subscription. The resource URI is that of a Metrics Advisor resource.
```

The resource URI must be that of the resource for which metrics are being queried. It's normally of the format `/subscriptions/<id>/resourceGroups/<rg-name>/providers/<source>/topics/<resource-name>`.

To find the resource URI:

1. Navigate to your resource's page in the Azure portal.
2. From the **Overview** blade, select the **JSON View** link.
3. In the resulting JSON, copy the value of the `id` property.

> ```

Reviewer: ```suggestion
```javascript
```

    throw new Error("METRICS_RESOURCE_ID must be set in the environment for this sample");
  }

  const result = await metricsQueryClient.getMetricDefinitions(metricsResourceId);

  for (const definition of result.definitions) {
    console.log(`Definition = ${definition.name}`);
  }

  const firstMetric = result.definitions[0];

  console.log(`Picking an example metric to query: ${firstMetric.name}`);

  const metricsResponse = await metricsQueryClient.queryMetrics(
    metricsResourceId,
    Durations.last5Minutes,
    {
      metricNames: [firstMetric.name!],
      interval: "PT1M"
    }
  );

  console.log(
    `Query cost: ${metricsResponse.cost}, interval: ${metricsResponse.interval}, time span: ${metricsResponse.timespan}`
  );

  const metrics: Metric[] = metricsResponse.metrics;
  console.log(`Metrics:`, JSON.stringify(metrics, undefined, 2));
}

main().catch((err) => {
  console.error("The sample encountered an error:", err);
  process.exit(1);
});
```

### Advanced scenarios

#### Query multiple workspaces

The same log query can be executed across multiple Log Analytics workspaces. In addition to the KQL query, the following parameters are required:

> - `workspace_id` - The first (primary) workspace ID.
> - `additional_workspaces` - A list of workspaces, excluding the workspace provided in the `workspace_id` parameter. The parameter's list items may consist of the following identifier formats:

Reviewer: ```suggestion
- `workspaceId` - The first (primary) workspace ID.
- `additionalWorkspaces` - A list of workspaces, excluding the workspace provided in the `workspaceId` parameter. The parameter's list items may consist of the following identifier formats:
```

  - Qualified workspace names
  - Workspace IDs
  - Azure resource IDs

For example, the following query executes in three workspaces:

> ```

Reviewer: ```suggestion
```javascript
```

      query: "AppRequests | take 2",
      timespan: "PT1H"
    }
  ];

  const result = await logsQueryClient.queryLogsBatch({
    queries: queriesBatch
  });

  if (result.results == null) {
    throw new Error("No response for query");
  }

  let i = 0;
  for (const response of result.results) {
    console.log(`Results for query with id: ${response.id}`);

    if (response.error) {
      console.log(`Query had errors:`, response.error);
    } else {
      if (response.tables == null) {
        console.log(`No results for query`);
      } else {
        console.log(
          `Printing results from query '${queriesBatch[i].query}' for '${queriesBatch[i].timespan}'`
        );

        for (const table of response.tables) {
          const columnHeaderString = table.columns
            .map((column) => `${column.name}(${column.type}) `)
            .join("| ");
          console.log(columnHeaderString);

          for (const row of table.rows) {
            const columnValuesString = row.map((columnValue) => `'${columnValue}' `).join("| ");
            console.log(columnValuesString);
          }
        }
      }
    }
    // next query
    i++;
  }
}
```

### Query metrics

> The following example gets metrics for an [Azure Metrics Advisor](https://docs.microsoft.com/azure/applied-ai-services/metrics-advisor/overview) subscription. The resource URI is that of a Metrics Advisor resource.

Reviewer 2: Fix link and link name to use Azure Monitor  instead of Metrics Advisor

Author: I think this is used as an example to get the resource URI. Python used eventgrid as the example.

}
```

### Batch logs query

> The following example demonstrates sending multiple queries at the same time using batch query API. The queries can be represented as a list of `BatchQuery` objects.

Reviewer: ```suggestion
The following example demonstrates sending multiple queries at the same time using the batch query API. The queries can be represented as a list of `BatchQuery` objects.
```

export async function main() {
  const tokenCredential = new DefaultAzureCredential();
  const metricsQueryClient = new MetricsQueryClient(tokenCredential);

  if (!metricsResourceId) {
    throw new Error("METRICS_RESOURCE_ID must be set in the environment for this sample");
  }

  const result = await metricsQueryClient.getMetricDefinitions(metricsResourceId);

  for (const definition of result.definitions) {
    console.log(`Definition = ${definition.name}`);
  }

  const firstMetric = result.definitions[0];

  console.log(`Picking an example metric to query: ${firstMetric.name}`);

  const metricsResponse = await metricsQueryClient.queryMetrics(
    metricsResourceId,
    Durations.last5Minutes,
    {
      metricNames: [firstMetric.name!],
      interval: "PT1M"
    }
  );

  console.log(
    `Query cost: ${metricsResponse.cost}, interval: ${metricsResponse.interval}, time span: ${metricsResponse.timespan}`
  );

  const metrics: Metric[] = metricsResponse.metrics;
  console.log(`Metrics:`, JSON.stringify(metrics, undefined, 2));
}

main().catch((err) => {
  console.error("The sample encountered an error:", err);
  process.exit(1);
});
```

### Handle metrics response

> The metrics query API returns a `QueryMetricsResult` object. The `QueryMetricsResult` object contains properties such as a list of `Metric`-typed objects, `interval`, `namespace`, and `timespan`. The `Metric` objects list can be accessed using the `metrics` param. Each `Metric` object in this list contains a list of `TimeSeriesElement` objects. Each `TimeSeriesElement` contains `data` and `metadataValues` properties. In visual form, the object hierarchy of the response resembles the following structure:

Reviewer: ```suggestion
The metrics query API returns a `QueryMetricsResult` object. The `QueryMetricsResult` object contains properties such as a list of `Metric`-typed objects, `interval`, `namespace`, and `timespan`. The `Metric` objects list can be accessed using the `metrics` property. Each `Metric` object in this list contains a list of `TimeSeriesElement` objects. Each `TimeSeriesElement` contains `data` and `metadataValues` properties. In visual form, the object hierarchy of the response resembles the following structure:
```

|---timespan
|---interval
|---namespace
|---resourceRegion
|---metrics (list of `Metric` objects)
    |---id
    |---type
    |---name
    |---unit
    |---displayDescription
    |---errorCode
    |---timeseries (list of `TimeSeriesElement` objects)
        |---metadataValues
        |---data (list of data points represented by `MetricValue` objects)
```

#### Example of handling response

```javascript
import { DefaultAzureCredential } from "@azure/identity";
import { Durations, Metric, MetricsQueryClient } from "@azure/monitor-query";
import * as dotenv from "dotenv";
dotenv.config();

const metricsResourceId = process.env.METRICS_RESOURCE_ID;
export async function main() {
  const tokenCredential = new DefaultAzureCredential();
  const metricsQueryClient = new MetricsQueryClient(tokenCredential);

  if (!metricsResourceId) {
    throw new Error("METRICS_RESOURCE_ID must be set in the environment for this sample");
  }

  console.log(`Picking an example metric to query: ${firstMetric.name}`);

  const metricsResponse = await metricsQueryClient.queryMetrics(
    metricsResourceId,
    Durations.last5Minutes,
    {
      metricNames: ["MatchedEventCount"],
      interval: "PT1M",

>       aggregations: ["Count"]

Author: ```suggestion
      aggregations: [AggregationType.Count]
```
